jupyterhub/tests/test_spawner.py

```diff
@@ -270,3 +273,90 @@ def stop():
 
         def poll():
             pass
+
+
+class MockAPITokenSpawner(MockSpawner):
+    """Mock Spawner that sets its own API token"""
+    use_this_api_token = None
+    will_resume = True
+    def start(self):
+        if self.use_this_api_token:
+            self.api_token = self.use_this_api_token
+        else:
+            self.use_this_api_token = self.api_token
+        return super().start()
+
+
+@pytest.mark.gen_test
+def test_spawner_reuse_api_token(db, app):
+    user = add_user(app.db, app, name='snoopy')
+    if user.running:
+        yield user.stop()
+    assert user.api_tokens == []
+    spawner = user._new_spawner(name='', spawner_class=MockAPITokenSpawner)
+    user.spawners[''] = spawner
+    assert user.spawner is spawner
+    # first start: gets a new API token
+    yield user.spawn()
+    api_token = spawner.api_token
+    found = orm.APIToken.find(app.db, api_token)
+    assert found
+    assert found.user.name == user.name
+    assert user.api_tokens == [found]
+    yield user.stop()
+    yield user.spawn()
+    # verify re-use of API token
+    assert spawner.api_token == api_token
+    # verify that a new token was not created
+    assert user.api_tokens == [found]
+
+
+@pytest.mark.gen_test
+def test_spawner_insert_api_token(db, app):
```

Comment:
@minrk this test was confusing to me

Perhaps add a docstring
`"""Token provided by spawner is not in the db. Insert token into db as a user-provided token."""`